daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala

```diff
@@ -367,3 +368,20 @@ final case class ECollision(
   override protected def prettyInternal: String =
     s"collision between ${entity1.pretty} and ${entity2.pretty}"
 }
+
+final case class EModuleVersionDependencies(
+    pkgId: PackageId,
+    pkgLangVersion: LanguageVersion,
+    depPkgId: PackageId,
+    dependencyLangVersion: LanguageVersion,
+) extends ValidationError {
+  import com.daml.lf.transaction.VersionTimeline.Implicits._
+
+  assert(pkgId != pkgLangVersion)
```

Comment:
I don’t understand this check? This shouldn’t even typecheck if Scala didn’t insist on screwing you over by allowing heretogenous equality.

Reply:
Equalities and dis equalities are inherited from java, i.e. they are untyped :(